### PR TITLE
http: support additional content type

### LINKED ISSUE #902: http: more human friendly request body
First of all, thank you for the great library.

We are using the [https://godoc.org/go.uber.org/zap#AtomicLevel.ServeHTTP](AtomicLevel.ServeHTTP) endpoint for setting the log levels dynamically.

For operators, this is a bit cumbersome, because they have to specify a valid JSON object.

By default, curl uses  `application/x-www-form-urlencoded` as the content type in a PUT request.
I think it would be nice to accept that as content type in `ServeHTTP` because it is a lot easier to type.
Following would be the command to set the logging level to debug:

```
curl -X PUT localhost:3000/log/level -d level=debug
```

If the content type is not `application/x-www-form-urlencoded` just keep the previous behavior, and expect a JSON payload.

If this is something that would be accepted, I'm be happy to work on this.
I'm just not sure if this qualifies as a breaking change or not.

### PR DESCRIPTION
**Pull request by Oncilla**
Support `application/x-www-form-urlencoded` as an additional content
type for `AtomicLevel.ServeHTTP`.

This is the default content type for `curl -X POST`.

With this change, interacting with the HTTP endpoint is a bit more
user friendly:

```
curl -X PUT localhost:8080/log/level -d level=debug
```

Additionally, the unit tests for the HTTP handler are transformed to
a table driven approach.

fixes #902

---
This change is [Reviewable](https://reviewable.io/reviews/uber-go/zap/903)

### CODE REVIEW

File: http_handler.go
Comment (abhinav) [inline http_handler.go:103]: Optionally, this can be shortened.
```suggestion
				err := l.UnmarshalText([]byte(lvl))
				return &l, err
```

File: http_handler_test.go
Comment (abhinav) [inline http_handler_test.go:44]: Thanks for making this a table test. If you wouldn't mind, we have a preference for slice-based table tests rather than maps:

```
tests :=  []struct {
  desc string
  method string
  contentType string
  ...
}{
  ...
}

for _, tt := range tests {
  ..
```

(See also https://github.com/uber-go/guide/blob/master/style.md#test-tables)

File: http_handler_test.go
Comment (abhinav) [inline http_handler_test.go:117]: can we name the test server srv or server? 

```suggestion
			server := httptest.NewServer(lvl)
			defer server.Close()

			req, err := http.NewRequest(test.Method, server.URL, strings.NewReader(test.Body))
```

File: http_handler_test.go
Comment (abhinav) [inline http_handler_test.go:109]: For table tests, consider renaming the test cases to `tt`.

  for _, tt := range tests {

File: http_handler.go
Comment (abhinav) [inline http_handler.go:121]: Do we want a JSON output even for non-JSON input?

Maybe this is okay and we can add a plain text output later if we need it and the client has an `Accepts: text/plain`?

CC @prashantv

**Review by abhinav — COMMENTED** [review]
Thanks for the change! The core logic looks mostly good. I have some suggestions around style.

**Comment by Oncilla** [comment]
Thanks for the swift feedback. I hope I get around to address the comments this week.

File: http_handler.go
Comment (Oncilla) [reply http_handler.go]: I prefer being explicit here. Especially after switching to value type.

It is not immediately obvious whether the returned value is pre- or post-UnmarshalText without knowing that part of the spec by hart. (TBH I would need to look that up)

File: http_handler.go
Comment (Oncilla) [reply http_handler.go]: If we want to return non-JSON in the response, I would definitively go for content type negotiation through the `Accepts` header.

In any case, I would vote for this to be part of a follow-up PR if it is desired.

File: http_handler_test.go
Comment (Oncilla) [reply http_handler_test.go]: Done

File: http_handler.go
Comment (abhinav) [inline http_handler.go:101]: Minor: It looks like these methods could be top-level functions since they don't do anything with the AtomicLevel. I'm gonna change that.

**Review by abhinav — APPROVED** [review]

**Comment by abhinav** [comment]
Sorry for the delay in reviewing. This looks great!
I'm going to get a second review on this before merging.
Thanks for the change!

**Review by prashantv — COMMENTED** [review]
Change looks good, although one change we can make to simplify, but also provide an easier way to set the level: use `ParseForm()` and `FormValue(..)` to get the level field if it exists. It would simplify the code, while also allowing URL parameters to be used. Thoughts?

**Comment by Oncilla** [comment]
@prashantv nice. Did not know that exists.

Should I also add tests for the URL params encoding?

**Comment by prashantv** [comment]
> Should I also add tests for the URL params encoding?

Yes, that would be great if you could add a test for passing the parameter in a URL (and mention it in the docs). Thanks!

**Comment by Oncilla** [comment]
Done.

**Review by prashantv — APPROVED** [review]
Thanks for the contribution!

**Pull request merged**
